File: backend/app/schemas/recommendations.py
from typing import List, Optional, Literal
from pydantic import BaseModel, Field
from app.schemas.ai_recipes import GeneratedRecipeIngredient

class MatchingIngredient(BaseModel):
    pantry_item_id: int
    pantry_item_name: str
    recipe_ingredient_name: str
    pantry_quantity: float
    pantry_unit: str
    required_quantity: float
    required_unit: str

class MissingIngredient(BaseModel):
    ingredient_name: str
    required_quantity: float
    required_unit: str

class ExpiringIngredient(BaseModel):
    pantry_item_id: int
    pantry_item_name: str
    expiration_date: str
    days_until_expiration: int

class RecommendedRecipe(BaseModel):
    recipe_id: int
    recipe_name: str
    estimated_calories: Optional[int]
    preparation_time_minutes: Optional[int]
    image_url: Optional[str]
    instructions: str
    matching_ingredients: List[MatchingIngredient]
    missing_ingredients: List[MissingIngredient]
    match_score: float = Field(..., description="Score from 0-1 indicating how well pantry items match recipe")
    expiring_ingredients_used: List[ExpiringIngredient]

class RecommendationFilters(BaseModel):
    max_preparation_time: Optional[int] = Field(None, description="Maximum preparation time in minutes")
    max_calories: Optional[int] = Field(None, description="Maximum calories per serving")
    max_missing_ingredients: Optional[int] = Field(None, description="Maximum number of missing ingredients allowed")

class RecommendationSort(BaseModel):
    sort_by: Literal["match_score", "preparation_time", "calories", "expiring_ingredients"] = Field(
        "match_score", 
        description="Field to sort by"
    )
    sort_order: Literal["asc", "desc"] = Field(
        "desc", 
        description="Sort order - ascending or descending"
    )

class RecommendationMetadata(BaseModel):
    total_recipes_analyzed: int
    total_before_filters: int
    total_after_filters: int
    applied_filters: RecommendationFilters
    applied_sort: RecommendationSort

class RecipeRecommendationsResponse(BaseModel):
    recommendations: List[RecommendedRecipe]
    total_pantry_items: int
    metadata: RecommendationMetadata
    message: Optional[str] = Field(None, description="Informative message if no recommendations found")

class RecommendationsRequest(BaseModel):
    filters: Optional[RecommendationFilters] = None
    sort: Optional[RecommendationSort] = None

File: frontend/lib/api/recommendations.ts
// API service for recommendations-related operations
import { getSession } from 'next-auth/react';

// Use localhost for browser requests (Docker internal URLs don't work from browser)
const API_BASE_URL = typeof window !== 'undefined' 
  ? 'http://localhost:8000'  // Browser: use localhost
  : process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'; // Server: use env var

export interface MatchingIngredient {
  pantry_item_id: number
  pantry_item_name: string
  recipe_ingredient_name: string
  pantry_quantity: number
  pantry_unit: string
  required_quantity: number
  required_unit: string
}

export interface MissingIngredient {
  ingredient_name: string
  required_quantity: number
  required_unit: string
}

export interface ExpiringIngredient {
  pantry_item_id: number
  pantry_item_name: string
  expiration_date: string
  days_until_expiration: number
}

export interface RecommendedRecipe {
  recipe_id: number
  recipe_name: string
  estimated_calories: number | null
  preparation_time_minutes: number | null
  image_url: string | null
  instructions: string
  matching_ingredients: MatchingIngredient[]
  missing_ingredients: MissingIngredient[]
  match_score: number
  expiring_ingredients_used: ExpiringIngredient[]
}

export interface RecommendationFilters {
  max_preparation_time?: number
  max_calories?: number
  max_missing_ingredients?: number
}

export interface RecommendationSort {
  sort_by: 'match_score' | 'preparation_time' | 'calories' | 'expiring_ingredients'
  sort_order: 'asc' | 'desc'
}

export interface RecommendationMetadata {
  total_recipes_analyzed: number
  total_before_filters: number
  total_after_filters: number
  applied_filters: RecommendationFilters
  applied_sort: RecommendationSort
}

export interface RecipeRecommendationsResponse {
  recommendations: RecommendedRecipe[]
  total_pantry_items: number
  metadata: RecommendationMetadata
  message?: string
}

export interface GetRecommendationsParams {
  max_preparation_time?: number
  max_calories?: number
  max_missing_ingredients?: number
  sort_by?: 'match_score' | 'preparation_time' | 'calories' | 'expiring_ingredients'
  sort_order?: 'asc' | 'desc'
  use_preferences?: boolean
}

class RecommendationsAPI {
  private async getAuthHeaders(): Promise<HeadersInit> {
    const session = await getSession();
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
    };
    
    if (session && (session as any).accessToken) {
      (headers as any).Authorization = `Bearer ${(session as any).accessToken}`;
    }
    
    return headers;
  }

  async getRecommendations(params: GetRecommendationsParams = {}): Promise<RecipeRecommendationsResponse> {
    const searchParams = new URLSearchParams();
    
    if (params.max_preparation_time !== undefined) {
      searchParams.append('max_preparation_time', params.max_preparation_time.toString());
    }
    if (params.max_calories !== undefined) {
      searchParams.append('max_calories', params.max_calories.toString());
    }
    if (params.max_missing_ingredients !== undefined) {
      searchParams.append('max_missing_ingredients', params.max_missing_ingredients.toString());
    }
    if (params.sort_by) {
      searchParams.append('sort_by', params.sort_by);
    }
    if (params.sort_order) {
      searchParams.append('sort_order', params.sort_order);
    }
    if (params.use_preferences !== undefined) {
      searchParams.append('use_preferences', params.use_preferences.toString());
    }

    const url = `${API_BASE_URL}/api/v1/recommendations${searchParams.toString() ? `?${searchParams.toString()}` : ''}`;
    
    console.log('Making recommendations request to:', url);
    
    const response = await fetch(url, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error('Recommendations API Error Response:', errorText);
      throw new Error(`Failed to fetch recommendations: ${response.status} ${response.statusText}. ${errorText}`);
    }

    const data = await response.json();
    console.log('Recommendations API Response:', data);
    return data;
  }
}

export const recommendationsAPI = new RecommendationsAPI();
